### PR TITLE
Fix opt-in dialog closing bug

### LINKED ISSUE #2622: Analytics submission can be blocked forever if opt-in dialog is forcibly closed due to parent closing
When the `OptInDialog` is closed, we release the lock, resuming any pending submission jobs. The places we release the lock are `okPressed()`, `cancelPressed()`, and `handelShellCloseEvent()`.

```java
  @Override
  protected void okPressed() {
    ...
    answeredSignal.release();
  }

  @Override
  protected void cancelPressed() {
    ...
    answeredSignal.release();
  }

  /**
   * When the dialog closes in other ways than pressing the buttons.
   */
  @Override
  protected void handleShellCloseEvent() {
    ...
    answeredSignal.release();
  }
```

I thought `handleShellCloseEvent()` covered all the other cases than the cases of the buttons pressed, but I just realized it doesn't. For example, if the dialog is closed because its parent dialog is closed, the method is not called.

So, I need to find another method to override that can cover any case of dialog closing. Unfortunately, there does not seem such a method, which is surprising.

### PR DESCRIPTION
**Pull request by chanseokoh**
Fixes #2622.

Misc: removed Javadocs that do not make sense as public description for the methods.

### CODE REVIEW

**Comment by chanseokoh** [comment]
Wait, let me see if I can improve the interface.

File: plugins/com.google.cloud.tools.eclipse.usagetracker/src/com/google/cloud/tools/eclipse/usagetracker/AnalyticsPingManager.java [hunk 1]
```diff
@@ -276,15 +277,16 @@ private void showOptInDialogIfNeeded(final Shell parentShell) throws Interrupted
     if (shouldShowOptInDialog()) {
       Display display = PlatformUI.getWorkbench().getDisplay();
 
-      final OptInDialog[] dialogHolder = new OptInDialog[1];
+      final Boolean[] optInAnswer = new Boolean[1];
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.usagetracker/src/com/google/cloud/tools/eclipse/usagetracker/AnalyticsPingManager.java]: any reason not to use a little b boolean here instead?

File: plugins/com.google.cloud.tools.eclipse.usagetracker.test/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialogTest.java [hunk 2]
```diff
@@ -17,60 +17,84 @@
...
-    assertFalse(getIsOptInYes());
+  public void testReturnCode_dialogClosed() {
+    scheduleClosingDialogAfterOpen(CloseAction.CLOSE_SHELL);
+    assertNotEquals(Window.OK, dialog.open());
   }
 
-  private boolean getIsOptInYes() throws InterruptedException {
-    final Boolean[] answerHolder = new Boolean[1];
-    Thread thread = new Thread(new Runnable() {
+  private enum CloseAction { PRESS_OK, PRESS_CANCEL, CLOSE_SHELL };
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.usagetracker.test/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialogTest.java]: Should you test for `dispose()` (`DISPOSE_SHELL`)?  I guess it doesn't matter now with your blocking-dialog pattern.

File: plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/ui/CompositeUtil.java [hunk 3]
```diff
@@ -33,6 +34,15 @@ public boolean apply(Control control) {
     });
   }
 
+  public static Button findButton(Composite composite, final String text) {
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/ui/CompositeUtil.java]: TBH we should minimize these types of methods and use SWTBot's helpers (e.g., `SWTBot#buttonWithLabel()`).

File: plugins/com.google.cloud.tools.eclipse.usagetracker.test/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialogTest.java [hunk 4]
```diff
@@ -17,60 +17,84 @@
...
+    shell.addFocusListener(new FocusAdapter() {
       @Override
-      public void run() {
-        try {
-          answerHolder[0] = dialog.isOptInYes();
-        } catch (InterruptedException ex) {}
+      public void focusGained(FocusEvent e) {
+        switch (closeAction) {
+          case PRESS_OK:
+            Button okButton = CompositeUtil.findButton(dialog.getShell(), "Share");
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.usagetracker.test/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialogTest.java]: Why not use SWTBot directly?

File: plugins/com.google.cloud.tools.eclipse.usagetracker/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialog.java [hunk 5]
```diff
@@ -36,18 +34,12 @@
  */
 public class OptInDialog extends Dialog {
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.usagetracker/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialog.java]: Would be worth adding javadoc to indicate that SHARE &rarr; `Window.OK` and DO_NOT_SHARE &rarr; `Window.CANCEL`.

File: plugins/com.google.cloud.tools.eclipse.usagetracker.test/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialogTest.java (hunk 4 above)
Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.usagetracker.test/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialogTest.java]: As we discussed, I couldn't make SWTBot find buttons with `buttonWithLabel()`.

File: plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/ui/CompositeUtil.java (hunk 3 above)
Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.test.util/src/com/google/cloud/tools/eclipse/test/util/ui/CompositeUtil.java]: Having trouble making `buttonWithLabel()`. Not sure why it doesn't work.

File: plugins/com.google.cloud.tools.eclipse.usagetracker.test/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialogTest.java (hunk 2 above)
Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.usagetracker.test/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialogTest.java]: Tried to simulate disposing the shell, but I faced some difficulty with it. Let's settle with `CLOSE_SHELL`.

File: plugins/com.google.cloud.tools.eclipse.usagetracker/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialog.java (hunk 5 above)
Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.usagetracker/src/com/google/cloud/tools/eclipse/usagetracker/OptInDialog.java]: Will add the Javadoc.

**Pull request merged**